src/adaptation/precond.jl

```diff
@@ -2,389 +2,227 @@
-    M :: T
-    # TODO: implement temporary `δ` as `WelfordCov`
-end
+Base.resize!(pc::UnitMassMatrix, θ::AbstractVecOrMat) = nothing
 
-WelfordVar(::Type{T}, sz::Union{Tuple{Int},Tuple{Int,Int}}) where {T<:AbstractFloat} = WelfordVar(0, zeros(T, sz), zeros(T, sz))
+reset!(::UnitMassMatrix) = nothing
 
-function reset!(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    wv.n = 0
-    wv.μ .= zero(T)
-    wv.M .= zero(T)
-end
+getM⁻¹(::UnitMassMatrix{T}) where {T} = LinearAlgebra.UniformScaling{T}(one(T))
 
-function add_sample!(wv::WelfordVar, s::AbstractVecOrMat)
-    wv.n += 1
-    @unpack μ, M, n = wv
-    for i in eachindex(s)
-        δ = s[i] - μ[i]
-        μ[i] += δ / n
-        M[i] += δ * (s[i] - μ[i])
-    end
-end
+adapt!(
+    ::UnitMassMatrix,
+    ::AbstractVecOrMat{<:AbstractFloat},
+    ::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+) = nothing
 
-# https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
-function get_var(wv::WelfordVar{<:AbstractVecOrMat{T}}) where {T<:AbstractFloat}
-    n, M = T(wv.n), wv.M
-    @assert n >= 2 "Cannot get covariance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M .+ T(1e-3) * (5 / (n + 5))
-end
+## Diagonal mass matrix adaptor
 
-## Covariance estimator
+abstract type VarEstimator{T} <: MassMatrixAdaptor end
 
-abstract type CovEstimator{T} end
+getM⁻¹(ve::VarEstimator) = ve.var
 
-# NOTE: This naive covariance estimator is used only in testing.
-mutable struct NaiveCov{T<:AbstractVector{<:AbstractVector{<:AbstractFloat}}} <: CovEstimator{T}
+Base.string(ve::VarEstimator) = string(getM⁻¹(ve))
+
+function adapt!(
+    ve::VarEstimator,
+    θ::AbstractVecOrMat{<:AbstractFloat},
+    α::AbstractScalarOrVec{<:AbstractFloat},
+    is_update::Bool=true
+)
+    resize!(ve, θ)
+    push!(ve, θ)
+    if ve.n >= ve.n_min && is_update
+        ve.var .= getest(ve)
+    end
+end
+
+# NOTE: this naive variance estimator is used only in testing
+mutable struct NaiveVar{E<:AbstractVecOrMat{<:AbstractFloat},T<:AbstractVector{E}} <: VarEstimator{T}
     n :: Int
     S :: T
 end
 
-NaiveCov(::Type{T}, sz::Union{Tuple{Int}}) where {T<:AbstractFloat} = NaiveCov(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int,Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Matrix{T}}())
 
-function add_sample!(nc::NaiveCov, s::AbstractVector)
-    nc.n += 1
-    push!(nc.S, s)
-end
+NaiveVar(sz::Tuple{Vararg{Int}}; kwargs...) = NaiveVar(Float64, sz; kwargs...)
 
-function get_cov(nc::NaiveCov)
-    @assert nc.n >= 2 "Cannot get covariance with only one sample"
-    return Statistics.cov(nc.S)
+function Base.push!(nv::NaiveVar, s::AbstractVecOrMat)
+    nv.n += 1
+    push!(nv.S, s)
 end
 
-function reset!(nc::NaiveCov{T}) where {T}
-    nc.n = 0
-    nc.S = T()
+function reset!(nv::NaiveVar{T}) where {T}
+    nv.n = 0
+    nv.S = T()
```

Comment:
It seems `nv` could be immutable (see above) if you use
```suggestion
    resize!(nv.S, 0)
```